File: android/devbricksx-java-compiler/src/main/java/com/dailystudio/devbricksx/compiler/utils/NameUtils.java
package com.dailystudio.devbricksx.compiler.utils;

public class NameUtils {

    public static String lowerCamelCaseName(String name) {
        if (TextUtils.isEmpty(name)) {
            return null;
        }

        if (Character.isLowerCase(name.charAt(0))) {
            return name;
        }

        StringBuilder builder = new StringBuilder(name);

        builder.setCharAt(0, Character.toLowerCase(name.charAt(0)));

        return builder.toString();
    }

    public static String underscoreCaseName(String name) {
        if (TextUtils.isEmpty(name)) {
            return null;
        }

        StringBuilder builder = new StringBuilder();
        char ch;
        for (int i = 0; i < name.length(); i++) {
            ch = name.charAt(i);
            if (!Character.isAlphabetic(ch)
                    || Character.isLowerCase(ch)) {
                builder.append(ch);
            } else {
                builder.append('_');
                builder.append(Character.toLowerCase(ch));
            }
        }

        return builder.toString();
    }

}

File: devbricksx-compiler/src/main/kotlin/com/dailystudio/devbricksx/ksp/helper/NameUtils.kt
package com.dailystudio.devbricksx.ksp.helper

import java.util.*


fun String.capitalizeName(): String {
    return if (isEmpty()) {
        this
    } else {
        replaceFirstChar {
            if (it.isLowerCase()) {
                it.titlecase(Locale.getDefault())
            } else {
                it.toString()
            }
        }
    }
}

fun String.lowerCamelCaseName(): String {
    if (isBlank()) {
        return this
    }

    val builder = StringBuilder(this)

    for (i in this.indices) {
        if (Character.isAlphabetic(this[i].code)) {
            builder.setCharAt(i, Character.toLowerCase(this[i]))
            break
        }
    }

    return builder.toString()
}

fun String.toVariableOrParamName(): String {
    return this@toVariableOrParamName.lowerCamelCaseName()
}

fun String.toVariableOrParamNameOfCollection(): String {
    return buildString {
        append(this@toVariableOrParamNameOfCollection.toVariableOrParamName())
        append('s')
    }
}

fun String.underscoreLowerCamelCaseName(): String {
    return buildString {
        append("_")
        append(this@underscoreLowerCamelCaseName.lowerCamelCaseName())
    }
}

fun String.kotlinGetterName(): String {
    if (!startsWith("get")) {
        return this
    }

    return this.removePrefix("get").lowerCamelCaseName()
}

fun String.kebabCaseName(): String {
    if (isBlank()) {
        return this
    }

    val builder = StringBuilder()

    for (i in this.indices) {
        val c = this[i]

        if (Character.isAlphabetic(c.toInt())
                && Character.isUpperCase(c)
                && i != 0) {

            builder.append("-")
        }

        builder.append(c.toLowerCase())
    }

    return builder.toString()
}

fun String.underlineCaseName(): String {
    if (isBlank()) {
        return this
    }

    val builder = StringBuilder()

    for (i in this.indices) {
        val c = this[i]

        if (Character.isAlphabetic(c.code)
                && Character.isUpperCase(c)
                && i != 0) {

            builder.append("_")
        }

        builder.append(c.lowercase())
    }

    return builder.toString()
}
